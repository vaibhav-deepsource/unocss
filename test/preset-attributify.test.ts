import { createGenerator } from '@unocss/core'
import presetUno from '@unocss/preset-uno'
import presetAttributify, { variantAttributify } from '@unocss/preset-attributify'
import { describe, expect, test } from 'vitest'

describe('attributify', () => {
  const fixture1 = `
<button
  sm="[color:red]"
  md="[--var:var(--another)]"
  lg="bg-blue-600"
  class="absolute fixed"
  bg="blue-400 hover:blue-500 dark:!blue-500 dark:hover:blue-600"
  text="sm white"
  flex="~ col"
  p="t-2"
  pt="2"
  border="rounded-xl"
  :font="foo > bar ? 'mono' : 'sans'"
  v-bind:p="y-2 x-4"
  border="2 rounded blue-200"
  un-children="m-auto"
  pt2 rounded-sm
  inline-block
  transform
  translate-x-100%
  translate-y-[10%]
  rotate-30
  rotate-60="" ma=""
  m='\`
  1 2
  3
\`'
>
  Button
</button>
`

  const fixture2 = `
<template>
  <div h-80 text-center flex flex-col align-center select-none all:transition-400>
    <input type="checkbox" peer mt-a>
    <div mb-a group peer-checked="text-4xl">
      <div
        font-100 text-4xl mb--3 p-10
        bg-gradient="to-r from-yellow-400 via-red-500 to-pink-500"
      >
        ~
      </div>
      <div text-5xl font-100 sm="bg-blue-600">
        unocss
      </div>
      <div op-20 font-200 mt-1 tracking-wider group-hover="text-teal-400 op-50">
        Re-imaging Atomic CSS
      </div>
    </div>
  </div>
  <div flex>
    <div ma inline-flex relative h-14>
      <input type="text" m-0 pt-4 px-4 text-true-gray-800 peer placeholder>
      <label absolute leading-1rem left-4 pointer-events-none text-gray-7 top="1/3" transition="200 linear"
        peer-not-placeholder-shown="-translate-y-4 scale-75 origin-top-left text-green-500"
        peer-focus="-translate-y-4 scale-75 origin-top-left text-green-500"
      >Experience now!</label>
    </div>
  </div>
</template>
`

  const uno = createGenerator({
    presets: [
      presetAttributify({ strict: true }),
      presetUno({ attributifyPseudo: true }),
    ],
  })

  test('extractor1', async() => {
    expect(await uno.applyExtractors(fixture1)).toMatchSnapshot()
  })

  test('extractor2', async() => {
    expect(await uno.applyExtractors(fixture2)).toMatchSnapshot()
  })

  test('variant', async() => {
    const variant = variantAttributify({
      prefix: 'un-',
      prefixedOnly: false,
      strict: true,
    })

    expect(Array.from(await uno.applyExtractors(fixture1) || [])
      .map((i) => {
        const r = variant(i)
        return typeof r === 'string' ? r : r ? r.matcher : r
      }))
      .toMatchSnapshot()
  })

  test('fixture1', async() => {
    const { css } = await uno.generate(fixture1)
    expect(css).toMatchSnapshot()
  })

  test('fixture2', async() => {
    const { css } = await uno.generate(fixture2)
    expect(css).toMatchSnapshot()
  })
})
