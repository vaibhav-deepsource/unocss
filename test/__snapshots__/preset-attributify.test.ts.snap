// Vitest Snapshot v1

exports[`attributify > extractor1 1`] = `
Set {
  "[sm~=\\"[color:red]\\"]",
  "[md~=\\"[--var:var(--another)]\\"]",
  "[lg~=\\"bg-blue-600\\"]",
  "absolute",
  "fixed",
  "[bg~=\\"blue-400\\"]",
  "[bg~=\\"hover:blue-500\\"]",
  "[bg~=\\"dark:!blue-500\\"]",
  "[bg~=\\"dark:hover:blue-600\\"]",
  "[text~=\\"sm\\"]",
  "[text~=\\"white\\"]",
  "[flex~=\\"~\\"]",
  "[flex~=\\"col\\"]",
  "[p~=\\"t-2\\"]",
  "[pt~=\\"2\\"]",
  "[border~=\\"rounded-xl\\"]",
  "[font~=\\"foo\\"]",
  "[font~=\\">\\"]",
  "[font~=\\"bar\\"]",
  "[font~=\\"?\\"]",
  "[font~=\\"mono\\"]",
  "[font~=\\":\\"]",
  "[font~=\\"sans\\"]",
  "[p~=\\"y-2\\"]",
  "[p~=\\"x-4\\"]",
  "[border~=\\"2\\"]",
  "[border~=\\"rounded\\"]",
  "[border~=\\"blue-200\\"]",
  "[un-children~=\\"m-auto\\"]",
  "[pt2=\\"\\"]",
  "[rounded-sm=\\"\\"]",
  "[inline-block=\\"\\"]",
  "[transform=\\"\\"]",
  "[translate-x-100%=\\"\\"]",
  "[translate-y-=\\"\\"]",
  "[%=\\"\\"]",
  "[rotate-30=\\"\\"]",
  "[rotate-60=\\"\\"]",
  "[ma=\\"\\"]",
  "[m~=\\"1\\"]",
  "[m~=\\"2\\"]",
  "[m~=\\"3\\"]",
}
`;

exports[`attributify > extractor2 1`] = `
Set {
  "[h-80=\\"\\"]",
  "[text-center=\\"\\"]",
  "[flex=\\"\\"]",
  "[flex-col=\\"\\"]",
  "[align-center=\\"\\"]",
  "[select-none=\\"\\"]",
  "[all:transition-400=\\"\\"]",
  "[type~=\\"checkbox\\"]",
  "[peer=\\"\\"]",
  "[mt-a=\\"\\"]",
  "[mb-a=\\"\\"]",
  "[group=\\"\\"]",
  "[peer-checked~=\\"text-4xl\\"]",
  "[font-100=\\"\\"]",
  "[text-4xl=\\"\\"]",
  "[mb--3=\\"\\"]",
  "[p-10=\\"\\"]",
  "[bg-gradient~=\\"to-r\\"]",
  "[bg-gradient~=\\"from-yellow-400\\"]",
  "[bg-gradient~=\\"via-red-500\\"]",
  "[bg-gradient~=\\"to-pink-500\\"]",
  "[text-5xl=\\"\\"]",
  "[sm~=\\"bg-blue-600\\"]",
  "[op-20=\\"\\"]",
  "[font-200=\\"\\"]",
  "[mt-1=\\"\\"]",
  "[tracking-wider=\\"\\"]",
  "[group-hover~=\\"text-teal-400\\"]",
  "[group-hover~=\\"op-50\\"]",
  "[ma=\\"\\"]",
  "[inline-flex=\\"\\"]",
  "[relative=\\"\\"]",
  "[h-14=\\"\\"]",
  "[type~=\\"text\\"]",
  "[m-0=\\"\\"]",
  "[pt-4=\\"\\"]",
  "[px-4=\\"\\"]",
  "[text-true-gray-800=\\"\\"]",
  "[placeholder=\\"\\"]",
  "[absolute=\\"\\"]",
  "[leading-1rem=\\"\\"]",
  "[left-4=\\"\\"]",
  "[pointer-events-none=\\"\\"]",
  "[text-gray-7=\\"\\"]",
  "[top~=\\"1/3\\"]",
  "[transition~=\\"200\\"]",
  "[transition~=\\"linear\\"]",
  "[peer-not-placeholder-shown~=\\"-translate-y-4\\"]",
  "[peer-not-placeholder-shown~=\\"scale-75\\"]",
  "[peer-not-placeholder-shown~=\\"origin-top-left\\"]",
  "[peer-not-placeholder-shown~=\\"text-green-500\\"]",
  "[peer-focus~=\\"-translate-y-4\\"]",
  "[peer-focus~=\\"scale-75\\"]",
  "[peer-focus~=\\"origin-top-left\\"]",
  "[peer-focus~=\\"text-green-500\\"]",
}
`;

exports[`attributify > fixture1 1`] = `
"/* layer: default */
.absolute{position:absolute;}
.fixed{position:fixed;}
[m~=\\"\\\\31 \\"]{margin:0.25rem;}
[m~=\\"\\\\32 \\"]{margin:0.5rem;}
[m~=\\"\\\\33 \\"]{margin:0.75rem;}
[ma=\\"\\"],
[un-children~=\\"m-auto\\"]>*{margin:auto;}
[inline-block=\\"\\"]{display:inline-block;}
[flex~=\\"\\\\~\\"]{display:flex;}
[flex~=\\"col\\"]{flex-direction:column;}
[translate-x-100\\\\%=\\"\\"],
[rotate-30=\\"\\"],
[rotate-60=\\"\\"],
[transform=\\"\\"]{--un-rotate:0;--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-scale-x:1;--un-scale-y:1;--un-scale-z:1;--un-skew-x:0;--un-skew-y:0;--un-translate-x:0;--un-translate-y:0;--un-translate-z:0;--un-transform:translateX(var(--un-translate-x)) translateY(var(--un-translate-y)) translateZ(var(--un-translate-z)) rotate(var(--un-rotate)) rotateX(var(--un-rotate-x)) rotateY(var(--un-rotate-y)) rotateZ(var(--un-rotate-z)) skewX(var(--un-skew-x)) skewY(var(--un-skew-y)) scaleX(var(--un-scale-x)) scaleY(var(--un-scale-y)) scaleZ(var(--un-scale-z));}
[translate-x-100\\\\%=\\"\\"]{--un-translate-x:100%;transform:var(--un-transform);}
[rotate-30=\\"\\"]{--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-rotate:30deg;transform:var(--un-transform);}
[rotate-60=\\"\\"]{--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-rotate:60deg;transform:var(--un-transform);}
[transform=\\"\\"]{transform:var(--un-transform);}
[border~=\\"\\\\32 \\"]{border-width:2px;border-style:solid;}
[border~=\\"blue-200\\"]{--un-border-opacity:1;border-color:rgba(191,219,254,var(--un-border-opacity));}
[border~=\\"rounded-xl\\"]{border-radius:0.75rem;}
[border~=\\"rounded\\"]{border-radius:0.25rem;}
[rounded-sm=\\"\\"]{border-radius:0.125rem;}
.dark [bg~=\\"dark\\\\:\\\\!blue-500\\"]{--un-bg-opacity:1 !important;background-color:rgba(59,130,246,var(--un-bg-opacity)) !important;}
.dark [bg~=\\"dark\\\\:hover\\\\:blue-600\\"]:hover{--un-bg-opacity:1;background-color:rgba(37,99,235,var(--un-bg-opacity));}
[bg~=\\"blue-400\\"]{--un-bg-opacity:1;background-color:rgba(96,165,250,var(--un-bg-opacity));}
[bg~=\\"hover\\\\:blue-500\\"]:hover{--un-bg-opacity:1;background-color:rgba(59,130,246,var(--un-bg-opacity));}
[p~=\\"x-4\\"]{padding-left:1rem;padding-right:1rem;}
[p~=\\"y-2\\"]{padding-top:0.5rem;padding-bottom:0.5rem;}
[p~=\\"t-2\\"],
[pt~=\\"\\\\32 \\"],
[pt2=\\"\\"]{padding-top:0.5rem;}
[font~=\\"mono\\"]{font-family:ui-monospace,SFMono-Regular,Menlo,Monaco,Consolas,\\"Liberation Mono\\",\\"Courier New\\",monospace;}
[font~=\\"sans\\"]{font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,\\"Segoe UI\\",Roboto,\\"Helvetica Neue\\",Arial,\\"Noto Sans\\",sans-serif,\\"Apple Color Emoji\\",\\"Segoe UI Emoji\\",\\"Segoe UI Symbol\\",\\"Noto Color Emoji\\";}
[text~=\\"sm\\"]{font-size:0.875rem;line-height:1.25rem;}
[text~=\\"white\\"]{--un-text-opacity:1;color:rgba(255,255,255,var(--un-text-opacity));}
@media (min-width: 640px){
[sm~=\\"\\\\[color\\\\:red\\\\]\\"]{color:red;}
}
@media (min-width: 768px){
[md~=\\"\\\\[--var\\\\:var\\\\(--another\\\\)\\\\]\\"]{--var:var(--another);}
}
@media (min-width: 1024px){
[lg~=\\"bg-blue-600\\"]{--un-bg-opacity:1;background-color:rgba(37,99,235,var(--un-bg-opacity));}
}"
`;

exports[`attributify > fixture2 1`] = `
"/* layer: default */
[pointer-events-none=\\"\\"]{pointer-events:none;}
[absolute=\\"\\"]{position:absolute;}
[relative=\\"\\"]{position:relative;}
[left-4=\\"\\"]{left:1rem;}
[top~=\\"\\\\31 \\\\/3\\"]{top:33.3333333333%;}
[m-0=\\"\\"]{margin:0rem;}
[ma=\\"\\"]{margin:auto;}
[mb--3=\\"\\"]{margin-bottom:-0.75rem;}
[mb-a=\\"\\"]{margin-bottom:auto;}
[mt-1=\\"\\"]{margin-top:0.25rem;}
[mt-a=\\"\\"]{margin-top:auto;}
[h-14=\\"\\"]{height:3.5rem;}
[h-80=\\"\\"]{height:20rem;}
[flex=\\"\\"]{display:flex;}
[inline-flex=\\"\\"]{display:inline-flex;}
[flex-col=\\"\\"]{flex-direction:column;}
[peer=\\"\\"]:focus~[peer-focus~=\\"origin-top-left\\"],
[peer=\\"\\"]:not(:placeholder-shown)~[peer-not-placeholder-shown~=\\"origin-top-left\\"]{transform-origin:top left;}
[peer=\\"\\"]:focus~[peer-focus~=\\"-translate-y-4\\"],
[peer=\\"\\"]:not(:placeholder-shown)~[peer-not-placeholder-shown~=\\"-translate-y-4\\"],
[peer=\\"\\"]:focus~[peer-focus~=\\"scale-75\\"],
[peer=\\"\\"]:not(:placeholder-shown)~[peer-not-placeholder-shown~=\\"scale-75\\"]{--un-rotate:0;--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-scale-x:1;--un-scale-y:1;--un-scale-z:1;--un-skew-x:0;--un-skew-y:0;--un-translate-x:0;--un-translate-y:0;--un-translate-z:0;--un-transform:translateX(var(--un-translate-x)) translateY(var(--un-translate-y)) translateZ(var(--un-translate-z)) rotate(var(--un-rotate)) rotateX(var(--un-rotate-x)) rotateY(var(--un-rotate-y)) rotateZ(var(--un-rotate-z)) skewX(var(--un-skew-x)) skewY(var(--un-skew-y)) scaleX(var(--un-scale-x)) scaleY(var(--un-scale-y)) scaleZ(var(--un-scale-z));}
[peer=\\"\\"]:focus~[peer-focus~=\\"-translate-y-4\\"],
[peer=\\"\\"]:not(:placeholder-shown)~[peer-not-placeholder-shown~=\\"-translate-y-4\\"]{--un-translate-y:-1rem;transform:var(--un-transform);}
[peer=\\"\\"]:focus~[peer-focus~=\\"scale-75\\"],
[peer=\\"\\"]:not(:placeholder-shown)~[peer-not-placeholder-shown~=\\"scale-75\\"]{--un-scale-x:0.75;--un-scale-y:0.75;transform:var(--un-transform);}
[select-none=\\"\\"]{user-select:none;}
[bg-gradient~=\\"from-yellow-400\\"]{--un-gradient-from:rgba(250,204,21, var(--un-from-opacity, 1));--un-gradient-stops:var(--un-gradient-from), var(--un-gradient-to, rgba(255, 255, 255, 0));}
[bg-gradient~=\\"to-pink-500\\"]{--un-gradient-to:rgba(236,72,153, var(--un-to-opacity, 1));}
[bg-gradient~=\\"via-red-500\\"]{--un-gradient-stops:var(--un-gradient-from), rgba(239,68,68, var(--un-via-opacity, 1)), var(--un-gradient-to, rgba(255, 255, 255, 0));}
[bg-gradient~=\\"to-r\\"]{--un-gradient-shape:to right;--un-gradient:var(--un-gradient-shape), var(--un-gradient-stops);background-image:linear-gradient(var(--un-gradient));}
[p-10=\\"\\"]{padding:2.5rem;}
[px-4=\\"\\"]{padding-left:1rem;padding-right:1rem;}
[pt-4=\\"\\"]{padding-top:1rem;}
[text-center=\\"\\"]{text-align:center;}
[peer=\\"\\"]:checked~[peer-checked~=\\"text-4xl\\"],
[text-4xl=\\"\\"]{font-size:2.25rem;line-height:2.5rem;}
[text-5xl=\\"\\"]{font-size:3rem;line-height:1;}
[font-100=\\"\\"]{font-weight:100;}
[font-200=\\"\\"]{font-weight:200;}
[leading-1rem=\\"\\"]{line-height:1rem;}
[tracking-wider=\\"\\"]{letter-spacing:0.05em;}
[group=\\"\\"]:hover [group-hover~=\\"text-teal-400\\"]{--un-text-opacity:1;color:rgba(45,212,191,var(--un-text-opacity));}
[peer=\\"\\"]:focus~[peer-focus~=\\"text-green-500\\"],
[peer=\\"\\"]:not(:placeholder-shown)~[peer-not-placeholder-shown~=\\"text-green-500\\"]{--un-text-opacity:1;color:rgba(34,197,94,var(--un-text-opacity));}
[text-gray-7=\\"\\"]{--un-text-opacity:1;color:rgba(55,65,81,var(--un-text-opacity));}
[text-true-gray-800=\\"\\"]{--un-text-opacity:1;color:rgba(38,38,38,var(--un-text-opacity));}
[group=\\"\\"]:hover [group-hover~=\\"op-50\\"]{opacity:0.5;}
[op-20=\\"\\"]{opacity:0.2;}
[all\\\\:transition-400=\\"\\"] *{transition-property:color,background-color,border-color,text-decoration-color,fill,stroke,opacity,box-shadow,transform,filter,backdrop-filter;transition-timing-function:cubic-bezier(0.4,0,0.2,1);transition-duration:400ms;}
[transition~=\\"\\\\32 00\\"]{transition-property:color,background-color,border-color,text-decoration-color,fill,stroke,opacity,box-shadow,transform,filter,backdrop-filter;transition-timing-function:cubic-bezier(0.4,0,0.2,1);transition-duration:200ms;}
@media (min-width: 640px){
[sm~=\\"bg-blue-600\\"]{--un-bg-opacity:1;background-color:rgba(37,99,235,var(--un-bg-opacity));}
}"
`;

exports[`attributify > variant 1`] = `
[
  "sm-[color:red]",
  "md-[--var:var(--another)]",
  "lg-bg-blue-600",
  undefined,
  undefined,
  "bg-blue-400",
  "hover:bg-blue-500",
  "dark:!bg-blue-500",
  "dark:hover:bg-blue-600",
  "text-sm",
  "text-white",
  "flex",
  "flex-col",
  "p-t-2",
  "pt-2",
  "border-rounded-xl",
  "font-foo",
  "font->",
  "font-bar",
  "font-?",
  "font-mono",
  "font-:",
  "font-sans",
  "p-y-2",
  "p-x-4",
  "border-2",
  "border-rounded",
  "border-blue-200",
  "children-m-auto",
  "pt2",
  "rounded-sm",
  "inline-block",
  "transform",
  "translate-x-100%",
  "translate-y-",
  "%",
  "rotate-30",
  "rotate-60",
  "ma",
  "m-1",
  "m-2",
  "m-3",
]
`;
